provisioner/terraform/tfparse/tfextract.go

```diff
@@ -88,16 +112,280 @@ func WorkspaceTags(ctx context.Context, logger slog.Logger, module *tfconfig.Mod
+//
+// file. It also ensures that any uses of the `coder_workspace_tags` data source only
+// reference the following data types:
+// 1. Static variables
+// 2. Template variables
+// 3. Coder parameters
+// Any other data types are not allowed, as their values cannot be known at
+// the time of template import.
+func WorkspaceTagDefaultsFromFile(ctx context.Context, logger slog.Logger, file []byte, mimetype string) (tags map[string]string, err error) {
+	module, cleanup, err := loadModuleFromFile(file, mimetype)
+	if err != nil {
+		return nil, xerrors.Errorf("load module from file: %w", err)
+	}
+	defer func() {
+		if err := cleanup(); err != nil {
+			logger.Error(ctx, "failed to clean up", slog.Error(err))
+		}
+	}()
+
+	parser := hclparse.NewParser()
+
+	// This only gets us the expressions. We need to evaluate them.
+	// Example: var.region -> "us"
+	tags, err = WorkspaceTags(ctx, logger, parser, module)
+	if err != nil {
+		return nil, xerrors.Errorf("extract workspace tags: %w", err)
+	}
+
+	// To evaluate the expressions, we need to load the default values for
+	// variables and parameters.
+	varsDefaults, err := loadVarsDefaults(ctx, logger, maps.Values(module.Variables))
+	if err != nil {
+		return nil, xerrors.Errorf("load variable defaults: %w", err)
+	}
+	paramsDefaults, err := loadParamsDefaults(ctx, logger, parser, maps.Values(module.DataResources))
+	if err != nil {
+		return nil, xerrors.Errorf("load parameter defaults: %w", err)
+	}
+
+	// Evaluate the tags expressions given the inputs.
+	// This will resolve any variables or parameters to their default
+	// values.
+	evalTags, err := EvalProvisionerTags(varsDefaults, paramsDefaults, tags)
+	if err != nil {
+		return nil, xerrors.Errorf("eval provisioner tags: %w", err)
+	}
+
+	// Ensure that none of the tag values are empty after evaluation.
+	for k, v := range evalTags {
+		if len(strings.TrimSpace(v)) > 0 {
+			continue
+		}
+		return nil, xerrors.Errorf("provisioner tag %q evaluated to an empty value, please set a default value", k)
+	}
+	return evalTags, nil
+}
+
+func loadModuleFromFile(file []byte, mimetype string) (module *tfconfig.Module, cleanup func() error, err error) {
+	// Create a temporary directory
+	cleanup = func() error { return nil } // no-op cleanup
+	tmpDir, err := os.MkdirTemp("", "tfparse-*")
+	if err != nil {
+		return nil, cleanup, xerrors.Errorf("create temp dir: %w", err)
+	}
+	cleanup = func() error { // real cleanup
+		return os.RemoveAll(tmpDir)
+	}
+
+	// Untar the file into the temporary directory
+	var rdr io.Reader
+	switch mimetype {
+	case "application/x-tar":
+		rdr = bytes.NewReader(file)
+	case "application/zip":
+		zr, err := zip.NewReader(bytes.NewReader(file), int64(len(file)))
+		if err != nil {
+			return nil, cleanup, xerrors.Errorf("read zip file: %w", err)
+		}
+		tarBytes, err := archive.CreateTarFromZip(zr, maxFileSizeBytes)
```

Comment:
double-check: do we have any large modules that would be a performance issue? Otherwise, it would be more convenient to a have a direct reader from ANY format to _bytes_.

Reply:
I don't think the modules would be stored in the database file itself?